Lux/src/image.hpp

```diff
@@ -0,0 +1,125 @@
+    // default constructor - creates empty "stub" image
+    image() : dim( { 0, 0 } ), bounds(), ipbounds( { 0, 0 }, { 0, 0 } ), fpbounds( ipbounds ),
+        mip_me( false ), mipped( false ), mip_utd( false ) {}     
+    // creates image of particular size 
+    image( const vec2i& dims ) 
+        :  dim( dims ), bounds( { -1.0, ( 1.0 * dim.y ) / dim.x }, { 1.0, ( -1.0 * dim.y ) / dim.x } ), ipbounds( { 0, 0 }, dim ), fpbounds( ipbounds ), 
+           mip_me( false ), mipped( false ), mip_utd( false ) {}     
+    image( const vec2i& dims, const bb2f& bb ) :  dim( dims ), bounds( bb ), ipbounds( { 0, 0 }, dim ), fpbounds( ipbounds ) ,
+        mip_me( false ), mipped( false ), mip_utd( false ) {}     
+    // copy constructor
+    image( const I& img ) : dim( img.dim ), bounds( img.bounds ), ipbounds( img.ipbounds ), fpbounds( ipbounds ), 
+        mip_me( img.mip_me ), mipped( img.mipped ), mip_utd( img.mip_utd ) 
+        {
+            std::copy( img.base.begin(), img.base.end(), back_inserter( base ) );
+            mip_it();
+        }     
+
+    void reset();                          // clear memory & set dimensions to zero (mip_me remembered)
+    void use_mip( bool m );
+    vec2i get_dim();
+    void set_dim( const vec2i& dims );
+    bb2f get_bounds();
+    void set_bounds( const bb2f& bb );
+    bool compare_dims( const I& img ); // returns true if images have same dimensions
+
+    // Sample base image
+    T index( const vec2i& vi, const image_extend& extend = SAMP_SINGLE );
+    T sample( const vec2f& v, const image_extend& extend = SAMP_SINGLE );           // quick and dirty sampling of nearest pixel value
+    T smooth_sample( const vec2f& v, const image_extend& extend = SAMP_SINGLE );    // linearly interpolated between corner values
+
+    // sample mip-map
+    T index( const vec2i& vi, const int& level, const image_extend& extend = SAMP_SINGLE );                
+    T sample( const vec2f& v, const int& level, const image_extend& extend = SAMP_SINGLE );           // quick and dirty sampling of nearest pixel value
+    T smooth_sample( const vec2f& v, const float& level, const image_extend& extend = SAMP_SINGLE );  // linearly interpolated between corner values and mip-map levels
+
+    // size modification functions
+    void resize( vec2i siz );
+    void crop( const bb2i& bb );
+    void circle_crop( float ramp_width = 0.0f );	// Colors black everything outside of a centered circle
+
+    // pixel modification functions
+    //void grayscale();
+    void fill( const T& c );
+
+    // masking
+    void apply_mask( const I& mask, I& result );
+
+    // rendering
+    void splat( 
+        const vec2f& center, 			// coordinates of splat center
+        const float& scale, 			// radius of splat
+        const float& theta, 			// rotation in degrees
+        const T& tint,				    // change the color of splat
+        const I& g 			            // image of the splat
+	);
+
+    // operators
+    I& operator += ( I& rhs );
```

Comment:
i don't understand why you need a const and non-const form of += where the rhs is sometimes const. i was under the impression the rhs is always const with += (and all other similar ops below)

Reply:
Yes, that is weird. I don't understand why I am getting errors if I make the image `const` - maybe something having to do with linalg.h. I have to make the vector and float versions `const` in order to do expressions like `b *= 1.5f` where beauty is an fimage.